File: backend/app/schemas/permissions.py
"""
Document permission schemas (Phase 16 — RESTRICTED grant management).
"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class GrantCreateRequest(BaseModel):
    """Body of POST /documents/{id}/permissions."""

    user_id: str = Field(..., description="Grantee user UUID (same organization)")
    permission_type: Literal["read", "write", "admin"] = "read"
    expires_at: datetime | None = Field(
        None, description="Optional expiry — NULL means the grant never expires"
    )


class GrantUpdateRequest(BaseModel):
    """Body of PATCH /documents/{id}/permissions/{user_id} (optional expiry edit)."""

    permission_type: Literal["read", "write", "admin"] | None = None
    expires_at: datetime | None = None


class PermissionGrantItem(BaseModel):
    """One grant row as exposed by the list endpoint."""

    id: str
    user_id: str
    grantee_email: str | None = None
    grantee_full_name: str | None = None
    permission_type: str
    expires_at: datetime | None = None
    created_at: datetime


class PermissionListResponse(BaseModel):
    document_id: str
    items: list[PermissionGrantItem]

File: frontend/src/lib/api/permissions.ts
/**
 * Document permission API (Phase 16 — RESTRICTED grant management).
 *
 * Explicit per-user access grants on documents. The document owner and
 * users with the `document:admin` permission may manage grants; everyone
 * else is rejected by the backend (403 same-org / 404 cross-org).
 */

import { del, get, post } from '@/lib/api/client'

export interface PermissionGrant {
  id: string
  user_id: string
  grantee_email: string | null
  grantee_full_name: string | null
  permission_type: 'read' | 'write' | 'admin'
  expires_at: string | null
  created_at: string
}

export interface PermissionListResponse {
  document_id: string
  items: PermissionGrant[]
}

export interface GrantCreateRequest {
  user_id: string
  permission_type?: 'read' | 'write' | 'admin'
  expires_at?: string | null
}

/** List explicit grants on a document (owner or document:admin). */
export async function listDocumentPermissions(
  documentId: string,
): Promise<PermissionListResponse> {
  return get<PermissionListResponse>(`/documents/${documentId}/permissions`)
}

/** Grant a user explicit access (409 if a grant already exists). */
export async function grantDocumentPermission(
  documentId: string,
  body: GrantCreateRequest,
): Promise<PermissionGrant> {
  return post<PermissionGrant>(`/documents/${documentId}/permissions`, body)
}

/** Revoke a user's explicit grant. */
export async function revokeDocumentPermission(
  documentId: string,
  userId: string,
): Promise<void> {
  await del<void>(`/documents/${documentId}/permissions/${userId}`)
}
